### PR TITLE
[Feature] Support MaskFormer(NeurIPS'2021) in MMSeg 1.x

### PR DESCRIPTION
**Pull request by MengzhangLI**
Add MaskFormer(NeurIPS'2021) in dev-1.x.
 
image

Paper: https://arxiv.org/abs/2107.06278
Repo: https://github.com/facebookresearch/MaskFormer

- [x] Align Inference Metric
- [x] Align Training Metric
- [x] Add Unit Test

## Results about Aligning Inference metric

(1) Using [maskformer_R50_bs16_160k](https://github.com/facebookresearch/MaskFormer/blob/main/configs/ade20k-150/maskformer_R50_bs16_160k.yaml), the inference metric has been aligned.

image

image

## Results about Aligning training metric
WIP.

### CODE REVIEW

File: mmseg/models/decode_heads/maskformer_head.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import List, Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmdet.models.dense_heads import MaskFormerHead as MMDET_MaskFormerHead
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/maskformer_head.py:7]: try:
   import 
except

File: configs/maskformer/maskformer_r50_512x512_80k_ade20k.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                    act_cfg=dict(type='ReLU', inplace=True),
+                    ffn_drop=0.1,
+                    dropout_layer=None,
+                    add_identity=True),
+                # the following parameter was not used,
+                # just make current api happy
+                feedforward_channels=2048,
+                operation_order=('self_attn', 'norm', 'cross_attn', 'norm',
+                                 'ffn', 'norm')),
+            init_cfg=None),
```
Comment (Li-Qingyun) [inline configs/maskformer/maskformer_r50_512x512_80k_ade20k.py:76]: Hello Mr. Li, my name is Qingyun and I participated in refactoring the DETR-like algorithm of mmdet. 
We have deconstructed DetrTransformerDecoder in the new algorithm to build with registry mechanism in open-mmlab/mmdetection#8763 (The PR has been merged into `refactor-detr` branch or official mmdetection, and will be merge to 3.x when other DETRs are refactored). 
I would like to remind you that the logic of building decoder and the config may need to be modified.

Comment (MengzhangLI) [reply configs/maskformer/maskformer_r50_512x512_80k_ade20k.py:76]: > Hello Mr. Li, my name is Qingyun and I participated in refactoring the DETR-like algorithm of mmdet. We have deconstructed the registry mechanism to build DetrTransformerDecoder in the new implementation in [open-mmlab/mmdetection#8763](https://github.com/open-mmlab/mmdetection/pull/8763) (The PR has been merged into `refactor-detr` branch of official mmdetection repo, and will be merge to 3.x when other DETRs are refactored). I would like to remind you that the logic of building decoder and the config may need to be modified.

Thank you qingyun, very useful information for us. ;)

File: configs/maskformer/maskformer_r50_512x512_80k_ade20k.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            # it will not filter masks whose score is smaller than 0.5 .
+            filter_low_score=False)),
+    # training and testing settings
+    train_cfg=dict(),
+    test_cfg=dict(mode='whole'),
+)
+
+# optimizer
+optimizer = dict(
+    type='AdamW', lr=0.0001, betas=(0.9, 0.999), weight_decay=0.0005)
```
Comment (MeowZheng) [inline configs/maskformer/maskformer_r50_512x512_80k_ade20k.py:127]: weight_decay need to be checked 
https://github.com/facebookresearch/MaskFormer/blob/da3e60d85fdeedcb31476b5edd7d328826ce56cc/configs/ade20k-150/Base-ADE20K-150.yaml#L25

File: configs/maskformer/maskformer_r50_512x512_80k_ade20k.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+optim_wrapper = dict(
+    _delete_=True,
+    type='OptimWrapper',
+    optimizer=optimizer,
+    clip_grad=None,
```
Comment (MeowZheng) [inline configs/maskformer/maskformer_r50_512x512_80k_ade20k.py:132]: clip grad need to be checked

https://github.com/facebookresearch/MaskFormer/blob/da3e60d85fdeedcb31476b5edd7d328826ce56cc/configs/ade20k-150/Base-ADE20K-150.yaml#L29-L33

File: configs/maskformer/maskformer_r50_512x512_80k_ade20k.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
+_base_ = [
+    '../_base_/datasets/ade20k.py', '../_base_/default_runtime.py',
+    '../_base_/schedules/schedule_80k.py'
```
Comment (MeowZheng) [inline configs/maskformer/maskformer_r50_512x512_80k_ade20k.py:3]: https://github.com/facebookresearch/MaskFormer/blob/da3e60d85fdeedcb31476b5edd7d328826ce56cc/configs/ade20k-150/Base-ADE20K-150.yaml#L22

80k or 160k

File: configs/maskformer/maskformer_swin-t_upernet_8xb2-160k_ade20k-512x512.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        norm_cfg=backbone_norm_cfg,
+        init_cfg=dict(type='Pretrained', checkpoint=checkpoint_file)),
+    decode_head=dict(
+        type='MaskFormerHead',
+        in_channels=[96, 192, 384, 768],  # pass to pixel_decoder inside
+    ))
+
+# optimizer
+optimizer = dict(
+    type='AdamW', lr=0.0005, betas=(0.9, 0.999), weight_decay=0.01)
```
Comment (MeowZheng) [inline configs/maskformer/maskformer_swin-t_upernet_8xb2-160k_ade20k-512x512.py:35]: might check base lr
https://github.com/facebookresearch/MaskFormer/blob/da3e60d85fdeedcb31476b5edd7d328826ce56cc/configs/ade20k-150/swin/maskformer_swin_tiny_bs16_160k.yaml#L17

**Review by MeowZheng — COMMENTED** [review]
might check both of 2 params
https://github.com/facebookresearch/MaskFormer/blob/da3e60d85fdeedcb31476b5edd7d328826ce56cc/configs/ade20k-150/swin/maskformer_swin_tiny_bs16_160k.yaml#L21-L22

File: mmseg/models/decode_heads/maskformer_head.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    See `Per-Pixel Classification is Not All You Need for Semantic Segmentation
+    <https://arxiv.org/pdf/2107.06278>`_ for details.
+    """
+
+    def __init__(self,
+                 num_classes: int = 150,
+                 align_corners: bool = False,
+                 ignore_index: int = 255,
+                 **kwargs) -> None:
+        super().__init__(**kwargs)
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/maskformer_head.py:34]:
```suggestion
                 pixel_decoder=...
                 **kwargs) -> None:
        if  'mmdet.' in pixel_decoder:
            pixel_decoder = pixel_decoder[6:] 
        super().__init__(pixel_decoder=pixel_decoder, **kwargs)
```

File: configs/maskformer/maskformer_r50-d32_8xb2-160k_ade20k-512x512.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+                        type='mmdet.DiceCost',
+                        weight=1.0,
+                        pred_act=True,
+                        eps=1.0)
+                ]),
+            sampler=dict(type='mmdet.MaskPseudoSampler')),
+        test_cfg=dict(
+            panoptic_on=False,
+            # For now, the dataset does not support
+            # evaluating semantic segmentation metric.
```
Comment (MeowZheng) [inline configs/maskformer/maskformer_r50-d32_8xb2-160k_ade20k-512x512.py:116]: what does it mean

File: mmseg/models/decode_heads/maskformer_head.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        self.align_corners = align_corners
+        self.out_channels = num_classes
+        self.ignore_index = ignore_index
+
+        feat_channels = kwargs['feat_channels']
+        self.cls_embed = nn.Linear(feat_channels, self.num_classes + 1)
+
+    def loss(self, x: Tuple[Tensor], batch_data_samples: SampleList,
+             train_cfg: ConfigType) -> dict:
+        batch_img_metas = []
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/maskformer_head.py:48]: modify loss as #2255

File: configs/maskformer/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+optim_wrapper = dict(
+    _delete_=True,
+    type='OptimWrapper',
+    optimizer=optimizer,
+    clip_grad=dict(max_norm=0.01, norm_type=2),
+    paramwise_cfg=dict(
+        custom_keys={
+            'backbone': dict(lr_mult=1.0),
+            'pos_embed': dict(decay_mult=0.),
+            'norm': dict(decay_mult=0.)
```
Comment (MeowZheng) [inline configs/maskformer/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512.py:44]: please check the specific key in original repo

File: mmseg/models/decode_heads/maskformer_head.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        self.align_corners = True
+        self.num_classes = num_classes
+        self.align_corners = align_corners
+        self.out_channels = num_classes
+        self.ignore_index = ignore_index
+
+        feat_channels = kwargs['feat_channels']
+        self.cls_embed = nn.Linear(feat_channels, self.num_classes + 1)
+
+    def _seg_data_to_instance_data(self, batch_data_samples: SampleList):
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/maskformer_head.py:46]: add docstring

File: mmseg/models/decode_heads/maskformer_head.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+            else:
+                gt_masks = torch.stack(masks).squeeze(1)
+
+            instance_data = InstanceData(
+                labels=gt_labels, masks=gt_masks.long())
+            batch_gt_instances.append(instance_data)
+        return batch_gt_instances, batch_img_metas
+
+    def loss(self, x: Tuple[Tensor], batch_data_samples: SampleList,
+             train_cfg: ConfigType) -> dict:
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/maskformer_head.py:81]: add docstring

File: mmseg/models/decode_heads/maskformer_head.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        batch_data_samples = []
+        for metainfo in batch_img_metas:
+            metainfo['batch_input_shape'] = metainfo['img_shape']
+            batch_data_samples.append(SegDataSample(metainfo=metainfo))
+        all_cls_scores, all_mask_preds = self(x, batch_data_samples)
+        mask_cls_results = all_cls_scores[-1]
+        mask_pred_results = all_mask_preds[-1]
+
+        # upsample masks
+        img_shape = batch_img_metas[0]['batch_input_shape']
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/maskformer_head.py:159]: why just use 'img_shape'
```suggestion
        img_shape = batch_img_metas[0]['img_shape']
```

File: mmseg/models/decode_heads/maskformer_head.py [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+            test_cfg (ConfigType): Test config.
+
+        Returns:
+            Tensor: A tensor of segmentation mask.
+        """
+
+        batch_data_samples = []
+        for metainfo in batch_img_metas:
+            metainfo['batch_input_shape'] = metainfo['img_shape']
+            batch_data_samples.append(SegDataSample(metainfo=metainfo))
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/maskformer_head.py:153]:
```suggestion
```

File: mmseg/models/decode_heads/maskformer_head.py [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+        all_cls_scores, all_mask_preds = self(x, batch_data_samples)
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/maskformer_head.py:154]: please add some comments '# forward function of decode head from mmdet needs batch_data_samples as inputs'

File: mmseg/models/decode_heads/maskformer_head.py [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+            mask_pred_results,
+            size=(img_shape[0], img_shape[1]),
+            mode='bilinear',
+            align_corners=False)
+
+        # semantic inference
+        cls_score = F.softmax(mask_cls_results, dim=-1)[..., :-1]
+        mask_pred = mask_pred_results.sigmoid()
+        seg_mask = torch.einsum('bqc,bqhw->bchw', cls_score, mask_pred)
+        return seg_mask
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/maskformer_head.py:170]:
```suggestion
        seg_logits = torch.einsum('bqc,bqhw->bchw', cls_score, mask_pred)
        return seg_logits
```

File: configs/maskformer/README.md [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+| MaskFormer | Swin-S    | 512x512   | 160000  | 5.33     | 26.98          | 49.36 | -             | [config](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/configs/maskformer/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512.py) | [model](https://download.openmmlab.com/mmsegmentation/v0.5/maskformer/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512_20221115_114710-5ab67e58.pth) \| [log](https://download.openmmlab.com/mmsegmentation/v0.5/maskformer/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512_20221115_114710.json) |
+
+Note:
+
+- All experiments of MaskFormer are implemented with 8 V100 (32G) GPUs with 2 samplers per GPU.
+- The results of MaskFormer are relatively not stable, variance of `R-101-D32` is 44.7 to 46.0, variance of `Swin-S` is 49.0 to 49.8.
+- The ResNet backbones utilized in MaskFormer models are standard `ResNet` rather than `ResNetV1c`.
+- `MultiScaleFlipAug` is not supported in MMSegmentation 1.x version yet, we would add "ms+flip" results as soon as possible.
```
Comment (MeowZheng) [inline configs/maskformer/README.md:60]:
```suggestion
- Test time augmentation is not supported in MMSegmentation 1.x version yet, we would add "ms+flip" results as soon as possible.
```

File: configs/maskformer/README.md [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+| MaskFormer | Swin-S    | 512x512   | 160000  | 5.33     | 26.98          | 49.36 | -             | [config](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/configs/maskformer/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512.py) | [model](https://download.openmmlab.com/mmsegmentation/v0.5/maskformer/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512_20221115_114710-5ab67e58.pth) \| [log](https://download.openmmlab.com/mmsegmentation/v0.5/maskformer/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512/maskformer_swin-s_upernet_8xb2-160k_ade20k-512x512_20221115_114710.json) |
+
+Note:
+
+- All experiments of MaskFormer are implemented with 8 V100 (32G) GPUs with 2 samplers per GPU.
+- The results of MaskFormer are relatively not stable, variance of `R-101-D32` is 44.7 to 46.0, variance of `Swin-S` is 49.0 to 49.8.
```
Comment (MeowZheng) [inline configs/maskformer/README.md:58]:
```suggestion
- The results of MaskFormer are relatively not stable.  The accuracy (mIoU) of model with `R-101-D32` is from 44.7 to 46.0, and with `Swin-S` is from 49.0 to 49.8.
```

**Review by MeowZheng — APPROVED** [review]
